[SYCL][Doc] Update ITT instrumentation docs

alexbatashev opened the pull request:
Add more details about ITT annotations implementation in SYCL runtime and update feature controls documentation.

bader commented on sycl/doc/extensions/ITTAnnotations/ITTAnnotations.rst:
```diff
@@ -65,7 +65,9 @@ a wrapper.
 
 Conditional compilation
 -----------------------
```
```suggestion
-----------------------

```

MrSidims commented on sycl/doc/extensions/ITTAnnotations/ITTAnnotations.rst:
```diff
@@ -66,6 +66,9 @@ a wrapper.
 -----------------------
 
+Data Parallel C++ compiler automatically instruments user code through
+SPIRITTAnnotations LLVM pass, which is enabled for targets, that natively
```
Note, that not everything is being annotated: just barriers, atomics, work item start/finish.

alexbatashev replied:
Thanks, I've added that bit.

gmlueck commented on sycl/doc/extensions/ITTAnnotations/ITTAnnotations.rst:
```diff
@@ -77,3 +81,7 @@ For this purpose we reserve a 1-byte specialization constant numbered
 in SYCL device code.
+
+The specialization constant value is controlled by
+INTEL_ENABLE_OFFLOAD_ANNOTATIONS environment variable. Tools, that support ITT
```
Should this environment variable be added to [EnvironmentVariables.md](https://github.com/intel/llvm/blob/sycl/sycl/doc/EnvironmentVariables.md)?  Note that all those environment variables start with the "SYCL_" prefix.  Should this one?

romanovvlad replied:
>> Should this environment variable be added to EnvironmentVariables.md? 

Not sure how it can be used outside of the tools that know about this var. But, since we check it, I agree that we should mention it in the EnvironmentVariables.md.

>> Note that all those environment variables start with the "SYCL_" prefix. Should this one?

This variable is used not only for the SYCL RT, that's why it has no SYCL_ prefix.

gmlueck reviewed (commented):
Is this document really a DPC++ extension specification?  It doesn't describe any API that the user can call from a DPC++ application.  Is it supposed to be a design document instead?  If so, maybe it should be moved outside of the "extensions" folder.

alexbatashev commented:
I moved the docs out of extension directory, converted it to markdown to match other design documents, added it to index page, and mentioned the env var on a special page.

pvchupin commented on sycl/doc/EnvironmentVariables.md:
```diff
@@ -47,6 +47,7 @@ subject to change. Do not rely on these variables in production code.
 | `SYCL_CACHE_THRESHOLD` | Positive integer | Cache eviction threshold in days (default value is 7 for 1 week). Set to 0 for disabling time-based cache eviction. |
 | `SYCL_CACHE_MIN_DEVICE_IMAGE_SIZE` | Positive integer | Minimum size of device code image in bytes which is reasonable to cache on disk because disk access operation may take more time than do JIT compilation for it. Default value is 0 to cache all images. |
 | `SYCL_CACHE_MAX_DEVICE_IMAGE_SIZE` | Positive integer | Maximum size of device image in bytes which is cached. Too big kernels may overload disk too fast. Default value is 1 GB. |
+| `INTEL_ENABLE_OFFLOAD_ANNOTATIONS` | Any(\*) | Enables ITT Annotations support for SYCL runtime. This variable should only be used by tools, that support ITT Annotations. |
```
Let's start splitting USER vars and special/experimental use vars and state clearly that these must not be used by end users.

alexbatashev replied:
I have created a separate file for env variables, that are not meant to be accessed by SYCL users. A separate file is needed so that we can put two distinct entries in https://intel.github.io/llvm-docs/. When this patch is merged, I plan to move rest of the debug variables from this file to the new one and move `Environment Variables` item to `Using oneAPI DPC++ for Application Development` section.

bader replied:
@pvchupin, I'm not sure why and how DPC++ should differentiate environment variables based on whom they are intended for "human" users or "tools". Could you clarify how do we define "end users"?

@alexbatashev, I think adding another table for environment variables intended for "tools" should resolve this comment and IMHO is better than adding a separate document.

pvchupin replied:
@bader, 
My comment was mostly about "experimental/internal-use-only" vs. "production", rather than "tools" vs. "human". 
I believe disclaimer we currently have is not working well anymore... some of these are used widely and some of these must never be used.
Separating these to different files sounds like a good idea. Would be good to split and categorize them clearly and draw the line more explicitly.
I would suggest to keep EnvironmentVariables.md (as it was referred many times in the past) and move the most of the variables to InternalEnvironmentVariables.md.

bader replied:
> My comment was mostly about "experimental/internal-use-only" vs. "production", rather than "tools" vs. "human".
> I believe disclaimer we currently have is not working well anymore... some of these are used widely and some of these must never be used.

I believe variables that "must never be used" must be removed.
I also suggest disabling by default (what I think you mean by "production") features which are "experimental/internal-use-only". There should be an option to enable those if needed, but this must be conscious action.

I don't see how moving around documentation can solve the usage problem. If feature is available, it's going to be used by someone eventually.

pvchupin replied:
Somebody added them, so I assume there was/is "experimental" reason at least. It's probably fine to disable by default or remove some, on the other hand there could be cases where experiment can be done randomly on existing build, e.g. during triaging.
I believe there are many vars now and categorizing them properly should help to at least some extent.

alexbatashev replied:
@pvchupin @bader I propose the following: I will move this variable back to EnvironmentVariables.md, and we will keep this discussion offline. I will prepare a followup patch once we have a decision.

pvchupin replied:
Ok. Thanks.

romanovvlad commented on sycl/doc/InternalEnvironmentVariables.md:
```diff
@@ -0,0 +1,12 @@
+# Internal Environment Variables
+
+**Warning:** the environment variables described in this document are used for
+development and debugging of DPC++ compiler and runtime. Their semantics are
```
```suggestion
development and debugging of DPC++ compiler and runtime or for interaction with other tools. Their semantics are
```

pvchupin commented:
@alexbatashev, please fill PR description with what's changed in this patch (to be used as commit message on merge).
